### ISSUE TITLE
Can’t run this kind of executable

### ISSUE THREAD

**Issue by Luckytrang2010** [holzschu/a-shell#293]
Lately, I wanted to test ngrok (port-forwarding stuff) so I decided to try it. Then I came across this.
![image](https://user-images.githubusercontent.com/53712708/128431985-c885e5a3-7ac7-4b82-aca2-820ee097f5bb.jpeg)

Although I'm aware it's an ELF and I'm obviously doing this on darwin.
![image](https://user-images.githubusercontent.com/53712708/128432130-4687b1e5-68fc-47f0-bba3-1296f7c19abe.jpeg)
_(I tried making ngrok a command)_
![image](https://user-images.githubusercontent.com/53712708/128432161-e49d52be-afdf-4e10-84ba-30b220ff4382.jpeg)![image](https://user-images.githubusercontent.com/53712708/128432162-6f04d735-bcda-452e-b7a7-75e3b9e6fa3d.jpeg)
![image](https://user-images.githubusercontent.com/53712708/128432167-ebc4c47d-d298-4583-98ee-c70c28a84a53.jpeg)

It led me to an suggestion:
Can you make the shell able to run executables like `./file_name`?

would be great for your response 👌

**Comment by holzschu** [comment]
iOS/iPadOS place multiple restrictions on executables and dynamic libraries: they must be compiled to Arm64 (the processor on iPhones and iPads), they must be signed with a developer certificate and they must be placed in `$APPDIR`. Anything that does not satisfy these three conditions will not be executed. To make things harder, `$APPDIR` can only be written at installation time.

Interpreted languages (Lua, Python, Perl) are accepted because, well, they are not compiled. WebAssembly binaries are accepted because Apple controls the WebAssembly JIT compiler, and also because it runs in a separate process, so it cannot access memory from the main app. 

Short version: to run a command in a-Shell, you need to either:
* port it to an interpreted language (Python?)
* compile it to WebAssembly, using the special wasi-sdk if it accesses files: https://github.com/holzschu/wasi-sdk
* compile it to Arm64, link it with the ios_system framework, embed it into the App and install your own version of the App: https://github.com/holzschu/ios_system/

Given that the source code of ngrok is not available, all three options are unavailable for this specific case. At this point, the most efficient path would be to ask ngrok developers whether they would consider creating a specific App.

**Comment by ifuchs** [comment]
FWIW, ngrok works fine under ish. I know this doesn’t help with a-shell as
ish and a-shell work in different ways but in case it helps the OP.

**Comment by dannytoomey** [comment]
Hello! I wanted to follow up on this thread and see if the above solution is still the simplest way to get `ngrok`, or any equivalent, to run on a-shell. I've tried running it on iSH with no luck, and I've tried other similar services with similar outcomes. I appreciate any insight or tips you have!

**Comment by holzschu** [comment]
I checked again, and  I don't see the source code for `ngrok` anywhere, so it cannot be cross-compiled.
Recent versions of a-Shell include an x86 interpreter (named `blink`) but it's unlikely to work, since `ngrok` probably requires many dynamic libraries.

That leaves open the question of "or any equivalent". I don't know of any equivalent command to `ngrok`, but you might have suggestions.

**Comment by dannytoomey** [comment]
Thanks for your reply! I dug around on this today and I found a [rust API](https://github.com/ngrok/ngrok-rust) for `ngrok` and a similar rust-based tool, `[tunnelto](https://github.com/agrinman/tunnelto)`. There's a good list of alternatives [here](https://github.com/anderspitman/awesome-tunneling). 

I was looking for rust tools to see if I could cross compile them to `wasi32-wasip1` and run them as a-Shell commands, though both of these projects appeared to have important dependencies that did not support this target. I'm thinking my best bet is probably to clone of one these repos and remove / replace dependencies that don't support `wasi32-wasip1`.

**Comment by holzschu** [comment]
It's highly likely that some of the Rust dependencies are related to sockets management, which doesn't exist in `wasm32-wasi*`. It could be added to the wasm engine we use, but the rust engine wouldn't know about it, and we'd have to rewrite the dependencies.
Since a-Shell has an integrated https server (because webworkers required it), and that server supports websockets, maybe the options that rely on websockets are the easiest path.
